// react-layout-kit/lib/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isVertical = exports.isSpaceDistribution = exports.isHorizontal = exports.getPrefix = exports.getFlexDirection = exports.getCssValue = void 0;
var getPrefix = function getPrefix(prefixCls) {
  if (prefixCls) return prefixCls;
  return 'layoutkit';
};
exports.getPrefix = getPrefix;
var getFlexDirection = function getFlexDirection(direction, isHorizontal) {
  if (isHorizontal) return 'row';
  switch (direction) {
    case 'horizontal':
      return 'row';
    case 'horizontal-reverse':
      return 'row-reverse';
    case 'vertical':
    default:
      return 'column';
    case 'vertical-reverse':
      return 'column-reverse';
  }
};
exports.getFlexDirection = getFlexDirection;
var isSpaceDistribution = function isSpaceDistribution(distribution) {
  if (!distribution) return;
  return ['space-between', 'space-around', 'space-evenly'].includes(distribution);
};
exports.isSpaceDistribution = isSpaceDistribution;
var isHorizontal = function isHorizontal(direction, _isHorizontal) {
  return getFlexDirection(direction, _isHorizontal) === 'row';
};
exports.isHorizontal = isHorizontal;
var isVertical = function isVertical(direction, isHorizontal) {
  return getFlexDirection(direction, isHorizontal) === 'column';
};
exports.isVertical = isVertical;
var getCssValue = function getCssValue(value) {
  return typeof value === 'number' ? "".concat(value, "px") : value;
};
exports.getCssValue = getCssValue;
