// @ant-design/pro-chat/es/ProChat/container/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["renderInputArea", "__PRO_CHAT_STORE_DEVTOOLS__", "showTitle", "style", "className", "chatItemRenderConfig", "backToBottomConfig", "appStyle"];
import { App as Container } from 'antd';
import App from "./App";
import { ProChatProvider } from "./Provider";
import { jsx as _jsx } from "react/jsx-runtime";
export function ProChat(_ref) {
  var renderInputArea = _ref.renderInputArea,
    __PRO_CHAT_STORE_DEVTOOLS__ = _ref.__PRO_CHAT_STORE_DEVTOOLS__,
    showTitle = _ref.showTitle,
    style = _ref.style,
    className = _ref.className,
    chatItemRenderConfig = _ref.chatItemRenderConfig,
    backToBottomConfig = _ref.backToBottomConfig,
    appStyle = _ref.appStyle,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/_jsx(ProChatProvider, _objectSpread(_objectSpread({}, props), {}, {
    devtoolOptions: __PRO_CHAT_STORE_DEVTOOLS__,
    children: /*#__PURE__*/_jsx(Container, {
      style: _objectSpread({
        height: '100%',
        width: '100%',
        position: 'relative'
      }, appStyle),
      className: className,
      children: /*#__PURE__*/_jsx(App, {
        chatItemRenderConfig: chatItemRenderConfig,
        renderInputArea: renderInputArea,
        chatRef: props.chatRef,
        showTitle: showTitle,
        style: style,
        backToBottomConfig: backToBottomConfig,
        className: className
      })
    })
  }));
}
