// @ant-design/pro-chat/es/ProChat/container/Provider.js
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["children", "devtoolOptions", "chats", "onChatsChange", "loading", "helloMessage", "userMeta", "assistantMeta", "request"];
import StoreUpdater from "./StoreUpdater";
import { memo } from 'react';
import { createStore as _createStore, Provider, useStoreApi } from "../store";
import { jsx as _jsx } from "react/jsx-runtime";
import { Fragment as _Fragment } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
export var ProChatProvider = /*#__PURE__*/memo(function (_ref) {
  var children = _ref.children,
    devtoolOptions = _ref.devtoolOptions,
    chats = _ref.chats,
    onChatsChange = _ref.onChatsChange,
    loading = _ref.loading,
    helloMessage = _ref.helloMessage,
    userMeta = _ref.userMeta,
    assistantMeta = _ref.assistantMeta,
    request = _ref.request,
    props = _objectWithoutProperties(_ref, _excluded);
  var isWrapped = true;
  var Content = /*#__PURE__*/_jsxs(_Fragment, {
    children: [children, /*#__PURE__*/_jsx(StoreUpdater, {
      chatRef: props.chatRef,
      init: !loading,
      helloMessage: helloMessage,
      chats: chats,
      userMeta: userMeta,
      request: request,
      assistantMeta: assistantMeta,
      onChatsChange: onChatsChange
    })]
  });
  try {
    useStoreApi();
  } catch (e) {
    isWrapped = false;
  }
  if (isWrapped) {
    return Content;
  }
  return /*#__PURE__*/_jsxs(Provider, {
    createStore: function createStore() {
      return _createStore(props, devtoolOptions);
    },
    children: [Content, /*#__PURE__*/_jsx(StoreUpdater, {
      chatRef: props.chatRef,
      init: !loading,
      helloMessage: helloMessage,
      chats: chats,
      userMeta: userMeta,
      request: request,
      assistantMeta: assistantMeta,
      onChatsChange: onChatsChange
    })]
  });
});
