// @ant-design/pro-chat/es/ProChat/hooks/useProChat.js
import { useMemo } from 'react';
import { useStoreApi } from "../store";
import { chatSelectors } from "../store/selectors";
import { useRefFunction } from "./useRefFunction";
export var useProChat = function useProChat() {
  var storeApi = useStoreApi();
  var _storeApi$getState = storeApi.getState(),
    resendMessage = _storeApi$getState.resendMessage,
    sendMessage = _storeApi$getState.sendMessage,
    stopGenerateMessage = _storeApi$getState.stopGenerateMessage,
    deleteMessage = _storeApi$getState.deleteMessage,
    clearMessage = _storeApi$getState.clearMessage,
    dispatchMessage = _storeApi$getState.dispatchMessage;
  var getChats = useRefFunction(function () {
    return storeApi.getState().chats;
  });
  var getChatMessages = useRefFunction(function () {
    return chatSelectors.currentChats(storeApi.getState());
  });
  var setMessageContent = useRefFunction(function (id, content) {
    dispatchMessage({
      type: 'updateMessage',
      id: id,
      key: 'content',
      value: content
    });
  });
  var setMessageValue = useRefFunction(function (id, key, value) {
    dispatchMessage({
      type: 'updateMessage',
      id: id,
      key: key,
      value: value
    });
  });
  return useMemo(function () {
    return {
      getChats: getChats,
      getChatMessages: getChatMessages,
      resendMessage: resendMessage,
      sendMessage: sendMessage,
      stopGenerateMessage: stopGenerateMessage,
      deleteMessage: deleteMessage,
      clearMessage: clearMessage,
      setMessageContent: setMessageContent,
      setMessageValue: setMessageValue
    };
  }, []);
};
